### DISCUSSION TITLE
use of --conflicts and -x

### DISCUSSION THREAD

**Discussion by johnrichardrinehart** [theory/sqitch#695] [Q&A] [answered]
It's not clear from the documentation how `sqitch add foo --conflicts bar` is used by `sqitch` when processing/generating the plan. I guess it means that the new change to the DB (`foo`) is expected to fail so long as `bar` is deployed. But, what's the point/benefit of annotating that? Does `sqitch` do any work to revert only the changes of `bar` such that `foo` will deploy successfully?

**Comment by theory** [comment, marked as answer]
Yes, the idea is that `foo` cannot be deployed as long as `bar` is also deployed. But you're right, it doesn't do anything to enforce it — or, rather, I don't think it does. The feature was never fully implemented, and thus is mostly informational right now. And also not very useful since there is no way to remove the `bar` change. I mean you can edit the plan, but what's the point of that?

In ten years no one has ever asked about this, or for the feature. So I don't think there's much demeaned for it — but all due credit to you for spotting it!

**Reply by johnrichardrinehart** [reply]
Ah, okay. I had similar questions around `requires`. Does `sqitch` use this annotation in the scripts to order execution (possibly during `rebase` operations) or flag/notify/signal/message about dependencies during execution?

"execution", here, is meant to refer to any of `deploy`, `revert`, `rebase`, etc. If that requires further clarification then please let me know.

I apologize - I know these answers are in the source code, but my `perl` is pretty awful.

**Comment by theory** [comment]
`requires` is actually used, though it's redundant if you have a single project, because order of deployment execution is exclusively determined by the plan file. The real power of `require` is requiring changes from other Sqitch projects. These are enforced Sqitch.

Note that order of reversion is determined not by the plan file, but the reverse order of deployment. So if you plan for changes `foo` and `bar`, then deploy, then add `fuzz` to the plan between `foo` and `bar`, running `sqitch rebase --onto `foo` will revert `bar`, then deploy `fuzz` and `bar`.

Perl, by the way, is AMAZING, I'll take no other opinions on the matter ;-P

**Reply by johnrichardrinehart** [reply]
> requires is actually used, though it's redundant if you have a single project, because order of deployment execution is exclusively determined by the plan file. The real power of require is requiring changes from other Sqitch projects. These are enforced Sqitch.

Thanks for explaining this! Can you please define "project" for me? I use `sqitch` currently to manage one `postgres` schema within one database. So, I'm confused how `sqitch` works in more general cases with either:
1) different database systems (MariaDB + CockroachDB + PostgreSQL, say)
2) multiple databases within the same DB system (so, `foo` and `bar` DBs both within 1 Postgres "instance")

> Note that order of reversion is determined not by the plan file, but the reverse order of deployment. So if you plan for changes foo and bar, then deploy, then add fuzz to the plan between foo and bar, running sqitch rebase --onto foowill revertbar, then deploy fuzzandbar`.

:exploding_head: :exploding_head: That's great to know! I'm working on getting my first round of migrations out the door for a release and I was struggling to insert changes after each deploy flow (I currently am developing doing something like `sqitch revert -y && sqitch deploy --verify` every time I finish a new `sqitch add`'s `deploy`, `verify`, and `revert`), since I don't have any `tag`s that I can `revert` to.

So, is it safe/recommended/normal/accepted to modify the plan file myself? If not, what does it mean to "add `fuzz` to the plan between `foo` and `bar`"?

> Perl, by the way, is AMAZING, I'll take no other opinions on the matter ;-P

:eyes: :eyes: I really appreciate `sqitch`......

**Comment by theory** [comment]
> Thanks for explaining this! Can you please define "project" for me? I use sqitch currently to manage one postgres schema within one database. So, I'm confused how sqitch works in more general cases with either:

Each plan file represents a project. The [sqitch configuration](https://sqitch.org/docs/manual/sqitch-configuration/) guide goes into detail on various patterns for maintaining multiple projects for different database engines and such.

> (I currently am developing doing something like sqitch revert -y && sqitch deploy --verify every time I finish a new sqitch add's deploy, verify, and revert), since I don't have any tags that I can revert to.

This is why `revert` exists, and `@HEAD` is always the meta-tag for the most recent change. So as I'm working on a change's scripts, I'm constantly rebasing it like so:

```
sqitch rebase --onto @HEAD^
```

Which reverts the most recent change, then applies whatever is in the plan from the change before that. The [sqitch changes](https://sqitch.org/docs/manual/sqitchchanges/) guide goes into some detail on how to specify changes and tags like this.

> So, is it safe/recommended/normal/accepted to modify the plan file myself? If not, what does it mean to "add fuzz to the plan between foo and bar"?

Once you're comfortable with the pattern, yes. This is why it's designed so that each change lives on exactly one line, so it's easy to reorder them in the file if you need to.

**Reply by johnrichardrinehart** [reply]
Thanks a lot!
